src/mmtk.rs

```diff
@@ -220,6 +220,56 @@ impl<VM: VMBinding> MMTK<VM> {
         }
     }
 
+    /// Initialize the GC worker threads that are required for doing garbage collections.
+    /// This is a mandatory call for a VM during its boot process once its thread system
+    /// is ready.  This call will invoke `Collection::spawn_gc_thread()`` to create GC threads.
+    ///
+    /// # Arguments
+    ///
+    /// *   `tls`: The thread that wants to enable the collection. This value will be passed back
+    ///     to the VM in `Collection::spawn_gc_thread()` so that the VM knows the context.
+    pub fn initialize_collection(&'static self, tls: VMThread) {
+        assert!(
+            !self.state.is_initialized(),
+            "MMTk collection has been initialized (was initialize_collection() already called before?)"
+        );
+        self.scheduler.spawn_gc_threads(self, tls);
+        self.state.initialized.store(true, Ordering::SeqCst);
+        probe!(mmtk, collection_initialized);
+    }
+
+    /// Prepare an MMTk instance for calling the `fork()` system call.
+    ///
+    /// This function makes all MMTk threads (currently including GC worker threads) save their
+    /// contexts and stop.  A subsequent call to `MMTK::after_fork()` will re-spawn the threads
+    /// using the saved contexts.
+    ///
+    /// This function returns when all MMTK threads stopped.
+    pub fn prepare_to_fork(&'static self) {
```

Comment:
`fork` is unix specific. We should either avoid using the term, or make it specific to certain targets. As we are not doing anything specific to forking in MMTk (we just stop GC threads), I think it may be a good idea to avoid using the term `fork`.

Reply:
Yes.  `fork` is unix-specific.  But I think we should be very explicit about forking because it has very unusual requirements.  For example, when calling `fork()`, the process must have only one thread.  If the process has multiple threads, the only thing safe to do is calling `exec()` or other async-signal-safe functions (see https://man7.org/linux/man-pages/man2/fork.2.html).  The implementation of `prepare_to_fork` is centered around those requirements.  The `GCThreadJoinHandle` trait which I just introduced yesterday is one example.   It waits until the underlying native thread of a GC thread quits.  It may be very different if we simply want to, for example, grow or shrink the number of GC threads at run time.  In that case, it is sufficient to let the GC thread give up their context, but the VM may repurpose those threads as other threads (such as used as mutators).

And the name should be exposed through the API.  If a VM needs forking (Android ART and CRuby), it knows precisely what it is doing.  CRuby carefully brings down other helper threads, too, before forking.

I am OK with making it platform-specific.  VMs treat forking as platform-specific, too.  CRuby only supports `Kernel#fork` if the underlying OS supports `fork()`, and Andorid only runs on its own modified Linux kernel.  But I don't know how to do it well now.  We need some kind of platform abstraction layer.  Then we can label platforms that support `fork()`.  That includes Linux, Mac, some UNIX variants, and CYGWIN, too.

Reply:
> Yes. `fork` is unix-specific. But I think we should be very explicit about forking because it has very unusual requirements. For example, when calling `fork()`, the process must have only one thread. If the process has multiple threads, the only thing safe to do is calling `exec()` or other async-signal-safe functions (see https://man7.org/linux/man-pages/man2/fork.2.html). The implementation of `prepare_to_fork` is centered around those requirements. The `GCThreadJoinHandle` trait which I just introduced yesterday is one example. It waits until the underlying native thread of a GC thread quits. It may be very different if we simply want to, for example, grow or shrink the number of GC threads at run time. In that case, it is sufficient to let the GC thread give up their context, but the VM may repurpose those threads as other threads (such as used as mutators).

In this API function, does MMTk core do anything specific to forking other than just stopping GC threads? Is it the binding rather than MMTk core that needs to know the semantics of forking?

Reply:
1.  It keeps the `GCWorker` struct instances so that GC workers can be re-spawn later, reusing those structs.  If we are adjusting the number of GC workers, we probably would drop those structs.  This probably doesn't matter because the GC workers are almost stateless when not running any work packets.
2.  It needs to ensure the underlying OS thread of the GC workers exit, not just returning from the function `memory_manager::start_worker`.  Of course the binding needs to cooperate by implementing the `GCThreadJoinHandle`.

There is an alternative design.  MMTk core provides a function to ask all GC threads to return from their entry-point functions, such as `memory_manager::start_worker`.  Then the VM bindings implement their own mechanisms to wait until their underlying OS threads exited.  By doing this, bindings no longer need to implement `GCThreadJoinHandle` and give it to mmtk-core.

Reply:
> 1. It keeps the `GCWorker` struct instances so that GC workers can be re-spawn later, reusing those structs.  If we are adjusting the number of GC workers, we probably would drop those structs.  This probably doesn't matter because the GC workers are almost stateless when not running any work packets.
> 2. It needs to ensure the underlying OS thread of the GC workers exit, not just returning from the function `memory_manager::start_worker`.  Of course the binding needs to cooperate by implementing the `GCThreadJoinHandle`.

It sounds to me that both are design choices rather than some requirements of `fork`.

Reply:
It is OK to destruct `GCWorker` when forking.  We will need to construct them again later.  It is fair to say this is a design choice.  It is an appropriate choice for forking.

It is OK to move the responsibility of waiting for the OS threads to the binding.  But either mmtk-core or the binding must wait for OS threads to exit.  This is a hard requirement for `fork`.